### PR TITLE
Add otlpmetric Temporality Preference env var

### PR DESCRIPTION
**Pull request by MadVikingGod**
This adds the `OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE` from the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk_exporters/otlp.md#additional-configuration)

### CODE REVIEW

**Review by hanyuancheung — APPROVED** [review]

**Comment by dmathieu** [comment]
Should there be tests? (and a changelog)

File: exporters/otlp/otlpmetric/internal/oconf/envconfig_test.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oconf
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
```
Comment (dmathieu) [inline exporters/otlp/otlpmetric/internal/oconf/envconfig_test.go:21]:
```suggestion
	"github.com/stretchr/testify/require"

```

**Review by dashpole — APPROVED** [review]

File: exporters/otlp/otlpmetric/internal/oconf/envconfig.go [hunk 2]
```diff
@@ -148,3 +151,38 @@ func withTLSConfig(c *tls.Config, fn func(*tls.Config)) func(e *envconfig.EnvOpt
...
+	return func(e *envconfig.EnvOptionsReader) {
+		if s, ok := e.GetEnvValue(n); ok {
+			switch strings.ToLower(s) {
+			case "cumulative":
+				fn(cumulativeTemporality)
+			case "delta":
+				fn(deltaTemporality)
+			case "lowmemory":
+				fn(lowMemory)
+			}
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/internal/oconf/envconfig.go:165]: Should the default log a warning about an unrecognized value?

File: exporters/otlp/otlpmetric/internal/oconf/envconfig.go [hunk 3]
```diff
@@ -148,3 +151,38 @@ func withTLSConfig(c *tls.Config, fn func(*tls.Config)) func(e *envconfig.EnvOpt
...
+		}
+	}
+}
+
+func cumulativeTemporality(metric.InstrumentKind) metricdata.Temporality {
+	return metricdata.CumulativeTemporality
+}
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/internal/oconf/envconfig.go:172]:
```suggestion
}

```

File: exporters/otlp/otlpmetric/internal/oconf/envconfig.go [hunk 4]
```diff
@@ -148,3 +151,38 @@ func withTLSConfig(c *tls.Config, fn func(*tls.Config)) func(e *envconfig.EnvOpt
...
+func deltaTemporality(ik metric.InstrumentKind) metricdata.Temporality {
+	switch ik {
+	case metric.InstrumentKindCounter, metric.InstrumentKindHistogram, metric.InstrumentKindObservableCounter:
+		return metricdata.DeltaTemporality
+	default:
+		return metricdata.CumulativeTemporality
+	}
+}
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/internal/oconf/envconfig.go:180]:
```suggestion
}

```

File: exporters/otlp/otlpmetric/internal/oconf/envconfig.go [hunk 5]
```diff
@@ -148,3 +152,42 @@ func withTLSConfig(c *tls.Config, fn func(*tls.Config)) func(e *envconfig.EnvOpt
...
+		if s, ok := e.GetEnvValue(n); ok {
+			switch strings.ToLower(s) {
+			case "cumulative":
+				fn(cumulativeTemporality)
+			case "delta":
+				fn(deltaTemporality)
+			case "lowmemory":
+				fn(lowMemory)
+			}
+			global.Warn("OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE is set to an invalid value. Defaulting to \"cumulative\".", "value", s)
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/internal/oconf/envconfig.go:167]: This is outside the `switch` statement, this warning will be logged for all set values of the environment variable.

File: exporters/otlp/otlpmetric/internal/oconf/envconfig.go [hunk 6]
```diff
@@ -148,3 +152,42 @@ func withTLSConfig(c *tls.Config, fn func(*tls.Config)) func(e *envconfig.EnvOpt
...
+}
+
+func deltaTemporality(ik metric.InstrumentKind) metricdata.Temporality {
+	switch ik {
+	case metric.InstrumentKindCounter, metric.InstrumentKindHistogram, metric.InstrumentKindObservableCounter:
+		return metricdata.DeltaTemporality
+	default:
+		return metricdata.CumulativeTemporality
+	}
+
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/internal/oconf/envconfig.go:183]:
```suggestion
```

**Comment by MadVikingGod** [comment]
[Comment from other PR](https://github.com/open-telemetry/opentelemetry-go/pull/4127#discussion_r1201541506): There is a lack of documentation around what Environment variables are supported. Could we create somewhere that users can find `OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE`?

**Comment by MadVikingGod** [comment]
For the documentation of the EnvVar there is a lack of documentation of the entire package.  I created a PR (#4336)  that includes Environment variables documentation.  @pellared can that satisfy the documentation ask here?

File: exporters/otlp/otlpmetric/internal/oconf/envconfig.go [hunk 7]
```diff
@@ -148,3 +152,42 @@ func withTLSConfig(c *tls.Config, fn func(*tls.Config)) func(e *envconfig.EnvOpt
...
+		if s, ok := e.GetEnvValue(n); ok {
+			switch strings.ToLower(s) {
+			case "cumulative":
+				fn(cumulativeTemporality)
+			case "delta":
+				fn(deltaTemporality)
+			case "lowmemory":
+				fn(lowMemory)
+			default:
+				global.Warn("OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE is set to an invalid value. Defaulting to \"cumulative\".", "value", s)
```
Comment (MrAlias) [inline exporters/otlp/otlpmetric/internal/oconf/envconfig.go:167]: This does not set the selector, like the test below suggests it shouldn't.

```suggestion
				global.Warn("OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE is set to an invalid value, ignoring.", "value", s)
```

**Review by MrAlias — APPROVED** [review]

**Review by pellared — APPROVED** [review]
> For the documentation of the EnvVar there is a lack of documentation of the entire package. I created a PR (#4336) that includes Environment variables documentation. @pellared can that satisfy the documentation ask here?

Sure 👍

**Pull request merged**